[Sema] Check exportability of @_spiOnly imported decls

xymus opened the pull request:
Extend the exportability checker to ensure that types and decls imported from an `@_spiOnly` import are only used in SPI signatures or implementation details, and not used in API. The check will always pick the less restrictive import between `@_spiOnly`, `@_implementationOnly` and the default import to apply the relevant check.

Plus, fix the leak allowing the use of implementation-only imported types in SPI by reporting it as an error when `-experimental-spi-only-imports` is set. This is the desired behavior when the feature will be completed.

This follows the module interface support for @_spiOnly imports in #60935. Next I'll integrate this feature with recent checks of API breakers (typealias desugaring in the swiftinterface, public use of conformances, and public imports of private modules), and report conflicting imports.

xymus committed:
[Sema] Check @_spiOnly imported decls exportability

Decls imported via an @_spiOnly import can only be used in @_spi decls
signatures, internal decl signatures and non-inlinable function bodies.

xymus committed:
[Sema] Error on implementation-only imported types in SPI with @_spiOnly

This used to be just a warning because it happened to work when clients
used only swiftmodules. Some projects have been using it as an
alternative to @_spiOnly imports. Now that we have an official feature
we can complete the diagnositc and error on implementation-only imports
types used in SPI.

xymus commented:
@swift-ci Please smoke test

tshortli commented on test/SPI/spi-only-import-exportability.swift:
```diff
@@ -0,0 +1,152 @@
+// RUN: %target-swift-emit-module-interface(%t/Client.swiftinterface) \
+// RUN:   %t/Client.swift -I %t -DSKIP_ERRORS \
+// RUN:   -experimental-spi-only-imports
+// RUN: %target-swift-typecheck-module-from-interface(%t/Client.swiftinterface) -I %t
```
Should we also generate and type check the private interface?

tshortli replied:
I had not thought about private interfaces when creating the `%target-swift-emit-module-interface` lit substitution; we may want another for convenience.

xymus replied:
Good point. Added it as `%target-swift-emit-module-interfaces(Main.swiftinterface, Main.private.swiftinterface)`. I kept the verify substitution as separate because it's likely to fail so I think it will be easier to debug that way.

xymus replied:
We could also keep only the current one `%target-swift-emit-module-interface` with a single argument and change the regex to then insert a `.private.` in the path. Do you have a preference for one or another?

tshortli replied:
I think I like the two argument version best because it's a little easier to piece together what's happening in the `RUN:` lines with that context, even if it is a bit more verbose.

xymus committed:
[Sema|lit] Intro lit substitution for generating the 2 swiftinterfaces

Intro the lit substitution
`%target-swift-emit-module-interfaces(Main.swiftinterface,
Main.private.swiftinterface` to generate both the public and private
swiftinterface. And use it in spi-only-import-exportability.swift.

To verify the generated interfaces I would recommend to keep two calls
to `%target-swift-typecheck-module-from-interface` as each is a
different invocation likely to fail.

xymus commented:
@swift-ci Please smoke test
